src/include/ref_inline.h

```diff
@@ -0,0 +1,147 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#pragma once
+
+/*
+ * __wt_ref_is_root --
+ *     Return if the page reference is for the root page.
+ */
+static WT_INLINE bool
+__wt_ref_is_root(WT_REF *ref)
+{
+    return (ref->home == NULL);
+}
+
+/*
+ * # The ref state API. #
+ *
+ * 5 functions* (3 macros) are defined to manipulate the ref state. This is a highly sensitive
+ * field and protected via the double underscore keyword. The field should only be accessed
+ * via these functions.
+ *
+ * __wt_ref_get_state:
+ * Get the state of the ref, wraps a relaxed atomic volatile load. At the time of writing this
+ * comment this was done to enable TSan and to enable burying the field behind the
+ * aforementioned double underscore.
+ *
+ * WT_REF_SET_STATE:
+ * Set the ref state. If HAVE_REF_TRACK is defined, track where the set call originated from. The
+ * ref state tracking is why we use macros here, since the tracking utilizes gcc identifiers to get
+ * the function and line number where the macro was called.
+ *
+ * WT_REF_CAS_STATE:
+ * Swap in a new state to the ref, tracking where the call originated from.
+ *
+ * __wt_ref_lock:
+ * Spin until the state WT_REF_LOCKED is swapped into the ref state field. Once the call to this
+ * function completes the caller has exclusive access to the ref.
+ *
+ * WT_REF_UNLOCK:
+ * Effectively wraps WT_REF_SET_STATE, however should only be used when returning the ref to the
+ * previous state as returned by __wt_ref_lock.
+ */
+
+/*
+ * __ref_set_state --
+ *     Set a ref's state. Accessed from the WT_REF_SET_STATE macro.
+ */
+static WT_INLINE void
+__ref_set_state(WT_REF *ref, uint8_t state)
+{
+    WT_RELEASE_WRITE_WITH_BARRIER(ref->__state, state);
+}
+
+#ifndef HAVE_REF_TRACK
+#define WT_REF_SET_STATE(ref, s) __ref_set_state((ref), (s))
+#else
+/*
+ * __ref_track_state --
+ *     Save tracking data when REF_TRACK is enabled. This is diagnostic code and we allow it to
```

Comment:
Great comment. Is it possible to clarify why "we allow it to race" is it because it is diagnostic code or for another reason?

Reply:
I believe it's because this is diagnostic code and some fuzziness in ref history is preferable to taking a lock and significant slowdowns. This was taken from the [original comment here](https://github.com/wiredtiger/wiredtiger/blob/3b8eb799ce99626cb8a827b0a8855440ba00fb24/src/include/btree_inline.h#L40).

We'll update this to `This is diagnostic code and ref->state changes are a hot path. As such we allow some racing in the history tracking code instead of requiring a lock and slowing down ref state transitions.`

Reply:
172bf5091